src/edu/stanford/nlp/tagger/maxent/TestSentence.java

```diff
@@ -574,60 +567,39 @@ private void cleanUpScorer() {
     return scores;
   }
 
-  // todo [cdm 2016]: Could this be sped up a bit by caching lambda array, extracting method for shared code?
-  // todo [cdm 2016]: Also it's allocating java.util.ArrayList$Itr for for loop - why can't it just random access array?
   /** Returns an unnormalized score (in log space) for each tag. */
   private double[] getApproximateHistories(String[] tags, History h, List<Pair<Integer,Extractor>> extractors, List<Pair<Integer,Extractor>> extractorsRare) {
```

Comment:
Any equivalent optimizations to be done in getExactHistories?

Reply:
Yes. Maybe it could be further unified; it seems the exact is going over all, the approximate only over a subset of tags.

Reply:
Thanks!  Maybe I'll change the names to runExactExtractor and runApproximateExtractor